File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if (head is None) or head.next is None:
            return False
        
        behind = head
        front = head
        while (behind.next is not None) and (front.next.next is not None):
            behind = behind.next
            front = front.next.next
            if behind == front:
                return True
            if (front is None) or (front.next is None):
                return False
        return False

File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) return false;
        ListNode behind = head;
        ListNode front = head;
        while ((behind.next != null) && (front.next.next != null)) {
            behind = behind.next;
            front = front.next.next;
            if (behind == front) return true;
            if ((front == null) || (front.next == null)) return false;
        }
        return false;
    }
}